rescheduler/rescheduler.go

```diff
@@ -77,8 +88,247 @@ func main() {
+	time.Sleep(10 * time.Minute)
+	podsBeingProcessed.Remove(pod)
+}
+
+func createKubeClient(flags *flag.FlagSet, inCluster bool) (*kube_client.Client, error) {
+	if inCluster {
+		return kube_client.NewInCluster()
+	}
+
+	clientConfig := kubectl_util.DefaultClientConfig(flags)
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		fmt.Errorf("error connecting to the client: %v", err)
+	}
+	return kube_client.NewOrDie(config), nil
+}
+
+func releaseAllTaints(client *kube_client.Client, nodeLister *kube_utils.ReadyNodeLister, podsBeingProcessed *podSet) {
+	nodes, err := nodeLister.List()
+	if err != nil {
+		glog.Warningf("Cannot release taints - error while listing nodes: %v", err)
+		return
+	}
+
+	for _, node := range nodes {
+		taints, err := kube_api.GetTaintsFromNodeAnnotations(node.Annotations)
+		if err != nil {
+			glog.Warningf("Error while getting Taints for node %v: %v", node.Name, err)
+			continue
+		}
+
+		newTaints := make([]kube_api.Taint, 0)
+		for _, taint := range taints {
+			if taint.Key == criticalAddonsOnly && podsBeingProcessed.HasId(taint.Value) {
+				glog.Infof("Releasing taint %+v on node %v", taint, node.Name)
+			} else {
+				newTaints = append(newTaints, taint)
+			}
+		}
+
+		if len(newTaints) != len(taints) {
+			taintsJson, err := json.Marshal(newTaints)
+			if err != nil {
+				glog.Warningf("Error while releasing taints on node %v: %v", node.Name, err)
+				continue
+			}
+
+			node.Annotations[kube_api.TaintsAnnotationKey] = string(taintsJson)
+			_, err = client.Nodes().Update(node)
+			if err != nil {
+				glog.Warningf("Error while releasing taints on node %v: %v", node.Name, err)
+			} else {
+				glog.Infof("Successfully released all taints on node %v", node.Name)
+			}
+		}
+	}
+}
+
+func prepareNodeForPod(client *kube_client.Client, predicateChecker *ca_simulator.PredicateChecker, node *kube_api.Node, criticalPod *kube_api.Pod) error {
+	// This taint will be removed if this function will return an error.
+	addTaint(client, node, podId(criticalPod))
+
+	requiredPods, otherPods, err := groupPods(client, node)
+	if err != nil {
+		return err
+	}
+
+	nodeInfo := schedulercache.NewNodeInfo(requiredPods...)
+	nodeInfo.SetNode(node)
+
+	// check whether critical pod still fit
+	if err := predicateChecker.CheckPredicates(criticalPod, nodeInfo); err != nil {
+		return fmt.Errorf("Pod %s doesn't fit to node %v: %v", podId(criticalPod), node.Name, err)
+	}
+	requiredPods = append(requiredPods, criticalPod)
+	nodeInfo = schedulercache.NewNodeInfo(requiredPods...)
+	nodeInfo.SetNode(node)
+
+	for _, p := range otherPods {
+		if err := predicateChecker.CheckPredicates(p, nodeInfo); err != nil {
+			glog.Infof("Pod %s will be deleted in order to schedule critical pods %s.", podId(p), podId(criticalPod))
+			// TODO(piosz): add better support of graceful deletion
+			delErr := client.Pods(p.Namespace).Delete(p.Name, kube_api.NewDeleteOptions(0))
```

Comment:
In the first version I wanted to not care about grace period at all, so changing to 10s which is reasonable small.